experimental/bucket4j-dynamodb-sdk-v1/src/main/java/io/github/bucket4j/dynamodb/v1/BaseDynamoDBTransaction.java

```diff
@@ -0,0 +1,94 @@
+package io.github.bucket4j.dynamodb.v1;
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
+import com.amazonaws.services.dynamodbv2.document.ItemUtils;
+import com.amazonaws.services.dynamodbv2.model.AttributeValue;
+import com.amazonaws.services.dynamodbv2.model.ConditionalCheckFailedException;
+import com.amazonaws.services.dynamodbv2.model.GetItemRequest;
+import com.amazonaws.services.dynamodbv2.model.PutItemRequest;
+import io.github.bucket4j.distributed.proxy.generic.compare_and_swap.CompareAndSwapBasedTransaction;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+abstract class BaseDynamoDBTransaction implements CompareAndSwapBasedTransaction {
+    private final AmazonDynamoDB db;
+    private final String table;
+    // NOTE: hardcode for now. Allow customization?
+    private final String keyAttr = Constants.Attrs.DEFAULT_KEY_NAME;
+    private final String stateAttr = Constants.Attrs.DEFAULT_STATE_NAME;
+
+    protected BaseDynamoDBTransaction(AmazonDynamoDB db, String table) {
+        this.db = Objects.requireNonNull(db, "DynamoDB is null");
+        this.table = Objects.requireNonNull(table, "table name is null");
+    }
+
+    @Override
+    public final Optional<byte[]> get() {
+        Map<String, AttributeValue> attrs = new HashMap<>();
+        attrs.put(keyAttr, getKeyAttributeValue());
+
+        GetItemRequest request = new GetItemRequest()
+                .withTableName(table)
+                .withKey(attrs)
+                .withConsistentRead(true);
+
+        Map<String, AttributeValue> result = db.getItem(request).getItem();
+        // per docs if item is not found, response will not contain item object
+        // at all. Based on this, check for emptiness seems redundant.
+        //
+        // docs: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_GetItem.html
+        // > If there is no matching item, GetItem does not return any data and there
+        // > will be no Item element in the response.
+        if (result == null || !result.containsKey(stateAttr)) {
+            return Optional.empty();
+        }
+
+        AttributeValue state = result.get(stateAttr);
+        if (state.getB() == null) {
```

Comment:
This is the first place I have a question for. So in my initial implementation I was returning empty optional here as I was not sure if throwing an exception from `get` is something that can be expected (method does not define if exception can be thrown if state is broken).

I switched to throwing as there is no possibility for code to succeed if state attribute has type different from `NULL` or `B`.

Reply:
This method does not support `NULL` value for `state` as from current logic `NULL` should never be written, but I can update code to support it.